File: 22-dp1/4-1d-dp-patterns/main.py
# 1D DP Patterns - Python example (Climbing Stairs)

def climb_stairs(n):
    if n <= 2:
        return n
    dp = [0] * (n+1)
    dp, dp = 1, 2

    for i in range(3, n+1):
        dp[i] = dp[i-1] + dp[i-2]

    return dp[n]

if __name__ == "__main__":
    print(climb_stairs(10))  # Output: 89

File: 22-dp1/4-1d-dp-patterns/main.cpp
// 1D DP Patterns - C++ example (Climbing Stairs)

#include <iostream>
#include <vector>

int climbStairs(int n) {
    if (n <= 2) return n;
    std::vector<int> dp(n+1);
    dp[1] = 1;
    dp = 2;

    for (int i = 3; i <= n; ++i)
        dp[i] = dp[i-1] + dp[i-2];

    return dp[n];
}

int main() {
    std::cout << climbStairs(10) << std::endl;  // Output: 89
    return 0;
}
